R/ncdf.R

```diff
@@ -419,25 +491,46 @@ read_ncdf = function(.x, ..., var = NULL, ncsub = NULL, curvilinear = character(
   var)
 }
 
-.get_data <- function(nc, var, dims, dimid_matcher) {
-  out_data <- lapply(var, function(.v) {
+.get_nc_var <- function(nc, .v, start, count) {
+  RNetCDF::var.get.nc(nc,
+                      variable = .v,
+                      start = start,
+                      count = count,
+                      collapse = FALSE, ## keep 1-dims
+                      unpack = TRUE,     ## offset and scale applied internally
+                      rawchar = TRUE)  ## needed for NC_CHAR, as per
+}
+
+.get_data <- function(nc, var, dims, dimid_matcher, pull = pull) {
+  out_data <- lapply(var, pull = pull, FUN = function(.v, pull) {
 
-    dm <- match(RNetCDF::var.inq.nc(nc, .v)$dimids,
-                dims[, "id", drop = TRUE])
-    if(is.null(dm)) dm <- c(1:nrow(dims))
+    if(is.null(dm <- dimid_matcher[.v][[1]])) dm <- seq_len(nrow(dims))
+
+    request <- list(start = dims[, "start", drop = TRUE][dm],
+                    count = dims[, "count", drop = TRUE][dm])
```

Comment:
note start and count will always be in canonical order for later subsetting purposes.